src/netex_idf/calendars.rs

```diff
@@ -24,30 +24,59 @@ use chrono::{Datelike, NaiveDateTime, Weekday};
 use failure::format_err;
 use log::warn;
 use minidom::Element;
-use std::collections::{BTreeSet, HashMap, HashSet};
+use std::{
+    cmp::{Ord, Ordering, PartialOrd},
+    collections::{BTreeSet, HashMap, HashSet},
+};
 
 type OperatingPeriods = HashMap<String, ValidityPeriod>;
-type DayTypeAssignments<'a> = HashMap<String, Vec<DayTypeAssignment<'a>>>;
+type DayTypeAssignments<'a> = HashMap<String, BTreeSet<DayTypeAssignment<'a>>>;
 pub type DayTypes = HashMap<String, BTreeSet<Date>>;
 
-#[derive(Debug)]
+#[derive(Debug, Eq, PartialEq)]
 enum DayTypeAssignment<'a> {
     OperatingPeriod(&'a ValidityPeriod),
     ActiveDay(Date),
     InactiveDay(Date),
 }
 
+// operating periods are treated first
+// then active days
+// then inactive days
+// to ensure that inactive days are removed if they appears before
+// operating periods in the XML feed.
+impl PartialOrd for DayTypeAssignment<'_> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        use DayTypeAssignment::*;
+        match (self, other) {
+            (OperatingPeriod(op1), OperatingPeriod(op2)) => {
+                op1.start_date.partial_cmp(&op2.start_date)
+            }
+            (ActiveDay(active1), ActiveDay(active2)) => active1.partial_cmp(&active2),
+            (OperatingPeriod(_), _) | (ActiveDay(_), _) => Some(Ordering::Less),
```

Comment:
This factorization is incorrect. It wouldn't be correct if `self` is of type `ActiveDay` and `other` is of type `OperatingPeriod`: you would expect `Ordering::Greater` but the actual code will return `Ordering::Less` which is a mistake.

```rust
#[test]  
// This test is failing                                                               
fn ord() {                                                              
    let dta1 = DayTypeAssignment::ActiveDay(Date::from_ymd(2020, 1, 1));
    let validity_period = ValidityPeriod {                              
        start_date: Date::from_ymd(2020, 1, 1),                         
        end_date: Date::from_ymd(2020, 1, 1),                           
    };                                                                  
    let dta2 = DayTypeAssignment::OperatingPeriod(&validity_period);    
    assert_eq!(Ordering::Greater, dta1.cmp(&dta2));                     
}                                                                       
```

The condition `(OperatingPeriod(_), _)` must be before `(ActiveDay(a1), ActiveDay(a2))`.

Reply:
fixed